### ISSUE TITLE
Only send modified records where possible.

### ISSUE THREAD

**Issue by thinkl33t** [pwaring/dns-sync-zone#35]
We've come across a bug with the way DNSAPI handle's ANAME records.  

In the recommended way of using the DNS API, they say:  
> This means that you can safely change a record by putting a DELETE followed by an ADD into the same HTTP request.

Unfortunately, when you do this to an ANAME record, even if it isn't changed, the related A record appears to be deleted, then recreated after a delay.  I suspect that DNSAPI is actually processing it as a deletion / insertion, not removing commands that don't do anything as the documentation would have you believe.

This has caused us a major outage, as 8.8.8.8 happened to hit the nameservers in between the record being deleted and recreated, and has cached the NXDOMAIN response.  

It would be nice if dns-sync-zone could take care of this, by comparing the current records, and in the provided zonefile, and only DELETE / ADD ing those which have changed.

**Comment by pwaring** [comment]
Sorry about that - I never use ANAME so had not come across this bug when I started writing the software.

Only sending modified records is probably a sensible thing to do anyway, just in case the API behaviour changes in future to not treat all commands as a single transaction (which would run the risk of silently breaking any records, not just ANAME).

**Comment by pwaring** [comment]
Relevant documentation for this feature is on the Mythic Beasts blog:

https://www.mythic-beasts.com/blog/2016/10/07/aname-records/

>  We resolve the ANAME and substitute the corresponding IP addresses into records into your zone. We then regularly check for any changes, and update your changes accordingly.

So it looks like if you create a new ANAME record there will always be a delay before the corresponding A and AAAA records are created.

What I haven't checked yet is whether the A and AAAA records generated by the ANAME are returned in the API LIST command. If they are I'm not sure how we would work out that those are linked to the ANAME and should be retained (i.e. not deleted because they are not in the zone file we are trying to synchronise).

**Comment by thinkl33t** [comment]
It happens! :man_shrugging: 

I was assuming from the docs there was some kind of cleverness happening on the mythic beasts side of the DNSAPI to only apply changes but apparently not :rofl:

**Comment by thinkl33t** [comment]
Ah!  The A records linked to an ANAME *are* being returned, and being deleted!

That explains a lot!

**Comment by thinkl33t** [comment]
I don't know if this is repeatable, but it looks like the autogenerated A records appear before the rest of the records, which are alphabetical.  We might just be able to crop out anything before the first @

**Comment by pwaring** [comment]
I've got a test domain hanging around for this sort of thing, so I'll try some tests tomorrow and see what's repeatable. Alternatively we might be able to just ignore any A and AAAA records for `@` if there's an ANAME.

**Comment by pwaring** [comment]
I've tested and confirmed a few things:

* The script will send DELETE commands for all A and AAAA records, even if they have been generated as the result of an ANAME record.
* There's a delay of about 5 minutes between uploading an ANAME record and the resulting A and AAAA records being generated.
* The LIST command appears to return results in this order: ns1/ns2 NS records, auto-generated A and AAAA record, other records (`@` first).

However, I don't know if the behaviour above is guaranteed or just happens to be the case because of the particular zone file I've uploaded. It's probably safest not to rely on it.

I think the safest thing to do is to handle records as follows:

 1. If a record exists in LIST and in the zone file, do nothing with that record.
 1. If a record exists in LIST and not in the zone file, DELETE it *unless* it is an A or AAAA record and there is an ANAME for the same hostname *or* it is an NS record for `@`.
 1. If a record exists in the zone file and not in LIST, upload it.

That should fix this particular bug, as well as making the software a bit more robust by performing a rsync-like synchronisation (i.e. only ADD/DELETE what has changed) rather than 'delete everything and then add all these records'.
